book/08-customizing-git/sections/hooks.asc

```diff
@@ -111,6 +111,14 @@ The first script to run when handling a push from a client is `pre-receive`.
 It takes a list of references that are being pushed from stdin; if it exits non-zero, none of them are accepted.
 You can use this hook to do things like make sure none of the updated references are non-fast-forwards, or to do access control for all the refs and files they're modifying with the push.
 
+===== `proc-receive`
+
+The `proc-receive` script is quite different from other server-side scripts, it is invoked by git push process, but can change the workflow or process between `pre-reveive` and `post-receive` hook.
+This's because if the upcoming update references match the specified reference pattern, which can be configured by `receive.procReceiveRefs`, then `proc-receive` will take over the whole flow
+after `pre-receive`, then ignore the `update` script. It can decide to create, update or rename new reference, or even fallback to previous process.
```

Comment:
"ignore the `update` script" is not the key point, we can ignore it here. We know that all the commands are fed to "pre-receive" hook in a batch, and each command to be executed by internal "exec_command()"  will be fed to the "update" hook, and the "proc-receive" hook should implement the "update" hook for other commands.